src/main/java/com/google/devtools/build/lib/bazel/bzlmod/SingleExtensionEvalFunction.java

```diff
@@ -322,11 +344,43 @@ private SingleExtensionEvalValue createSingleExtentionValue(
                     Function.identity())));
   }
 
-  private void validateAllImportsAreGenerated(
+  /**
+   * Validates the result of the module extension evaluation against the declared imports, throwing
+   * an exception if validation fails.
+   *
+   * <p>Since extension evaluation does not depend on the declared imports, the result of the
+   * evaluation can be reused and persisted in the lockfile even if validation fails.
+   */
+  private void validateExtensionResult(
       ImmutableMap<String, RepoSpec> generatedRepoSpecs,
+      Optional<ModuleExtensionMetadata> moduleExtensionMetadata,
+      ModuleExtensionId extensionId,
       SingleExtensionUsagesValue usagesValue,
-      ModuleExtensionId extensionId)
+      Environment env)
       throws SingleExtensionEvalFunctionException {
+    // Evaluate the metadata before failing on invalid imports so that fixup warning are still
+    // emitted in case of an error.
+    if (moduleExtensionMetadata.isPresent()) {
+      try {
+        moduleExtensionMetadata
+            .get()
+            .evaluate(
+                usagesValue.getExtensionUsages().values(),
+                generatedRepoSpecs.keySet(),
+                env.getListener());
+      } catch (EvalException e) {
```

Comment:
I think ideally we'd throw ExternalDepsException in ModuelExtensionMetadata#evaluate, since it was EvalException just for convenience. But doesn't need to be done in this PR.

Reply:
Agreed that this would be better, but I would prefer to do this in a separate PR.